### PR TITLE
Fix: add % method to BigDecimal

### LINKED ISSUE #9782: Undefined method '%' for BigDecimal
The `BigDecimal` class, in `big/big_decimal.cr`, has no `%` method, making it impossible to modulo a `BigDecimal`. This can be seen in this code:
```cr
require "big/big_decimal.cr"

a = BigDecimal.new 5
puts a % 2
```
Expected output: `1`
Actual output: `Error: undefined method % for BigDecimal`
This can be seen on [Carcin](https://carc.in/#/r/9rq6).

### PR DESCRIPTION
**Pull request by sugarfi**
This PR adds a `%` method to the `BigDecimal` class in `src/big/big_decimal.cr`. Should fix #9782 .

### CODE REVIEW

**Review by asterite — CHANGES_REQUESTED** [review]
Looks great! Could you add some tests?

**Comment by sugarfi** [comment]
> Could you add some tests?

Sure! I assume I'd just add some files in the spec directory?

**Comment by sugarfi** [comment]
All done, should have some specs now!

**Comment by asterite** [comment]
Thanks!

A couple of things:
- We usually write `some_computation.should eq(expected_value)` but in the spec you wrote it's the other way around
- It would be nice to have a few more specs to make sure all of the `if` branches you wrote are covered

**Comment by sugarfi** [comment]
> We usually write some_computation.should eq(expected_value) but in the spec you wrote it's the other way around

My bad, first time contributing. I'll go ahead and fix that!

**Comment by sugarfi** [comment]
I was looking at the BigDecimal specs though, and most of them are written similar to this:
```cr
BigDecimal.new(1).should eq(BigDecimal.new(1) / BigDecimal.new(1))
```

**Comment by asterite** [comment]
Yeah, I think I was absent when BigDecimal was introduced, the specs could have been written in a better way. For starters, we never have a huge single spec to test different methods. There's so much to improve in this class.

Since I don't have knowledge of this and I have very little time now, I'll leave others to review this.

Thank you again!

**Review by asterite — APPROVED** [review]
Looks good!

It might be nice to also have some tests that check that decimals are properly divided. For example checking the remainder of 5678.912 % 12.34. After all these are decimal types :-)

**Comment by bcardiff** [comment]
Thanks @sugarfi , I think that we need a couple of more specs before merging this to ensure things works as in float.

In `float_spec.cr`, the following specs are used to check for the result sign on negative operands, and some cases with decimal values.

```crystal
  describe "modulo" do
    it "raises when mods by zero" do
      expect_raises(DivisionByZeroError) { 1.2.modulo 0.0 }
    end

    it { (13.0.modulo 4.0).should eq(1.0) }
    it { (13.0.modulo -4.0).should eq(-3.0) }
    it { (-13.0.modulo 4.0).should eq(3.0) }
    it { (-13.0.modulo -4.0).should eq(-1.0) }
    it { (11.5.modulo 4.0).should eq(3.5) }
    it { (11.5.modulo -4.0).should eq(-0.5) }
    it { (-11.5.modulo 4.0).should eq(0.5) }
    it { (-11.5.modulo -4.0).should eq(-3.5) }
  end
```

The same test cases in BigDecimal would be enough.

float also has `#modulo` (as an alias to `%`) and `#reminder` with a different treatment of the sign. Those can be handled in another PR or include it here if wanted.

**Comment by sugarfi** [comment]
@bcardiff ok, will update the spec. I did implement this modulo using the built-in `%` operator - will that cause floating point noise issues?

**Comment by sugarfi** [comment]
These are the specs I just added, anything I need to change before I test and commit?
```cr
it "handles modulus correctly" do
    (BigDecimal.new(13.0) % BigDecimal.new(4.0)).should eq(BigDecimal.new(1.0))
    (BigDecimal.new(13.0) % BigDecimal.new(-4.0)).should eq(BigDecimal.new(-3.0))
    (BigDecimal.new(-13.0) % BigDecimal.new(4.0)).should eq(BigDecimal.new(3.0))
    (BigDecimal.new(-13.0) % BigDecimal.new(-4.0)).should eq(BigDecimal.new(-1.0))
    (BigDecimal.new(11.5) % BigDecimal.new(4.0)).should eq(BigDecimal.new(3.5))
    (BigDecimal.new(11.5) % BigDecimal.new(-4.0)).should eq(BigDecimal.new(-0.5))
    (BigDecimal.new(-11.5) % BigDecimal.new(4.0)).should eq(BigDecimal.new(0.5))
    (BigDecimal.new(-11.5) % BigDecimal.new(-4.0)).should eq(BigDecimal.new(-3.5))
end
```

**Comment by bcardiff** [comment]
The `DivisionByZeroError` one should be added also. 

I think they will all work directly thanks to the implementation, but better be safe.

I don't follow why you said: will that cause floating point noise issues?. The BigDecimal % implementation uses BigInt one. There are no floats involved I think.

File: src/big/big_decimal.cr [hunk 1]
```diff
@@ -189,6 +189,22 @@ struct BigDecimal < Number
...
+      BigDecimal.new(@value % scaled.value, @scale)
+    elsif @scale < other.scale
+      scaled = scale_to(other)
+      BigDecimal.new(scaled.value % other.value, other.scale)
+    else
+      BigDecimal.new(@value % other.value, @scale)
+    end
+  end
+
+  def %(other : Int)
```
Comment (Sija) [inline src/big/big_decimal.cr:204]: Same overload could be added for `Float` perhaps?

Comment (sugarfi) [reply src/big/big_decimal.cr:204]: Yeah, I'll add that. Thanks!

**Comment by sugarfi** [comment]
@bcardiff @Sija ok, I've committed both of the changes you suggested: the spec handles `DivisionByZeroError`, and there is an overload `BigDecimal#%(other : Float)`.

**Comment by sugarfi** [comment]
Just realized that `BigDecimal#**(BigDecimal)` isn't a thing, so I added a fix and some quick specs for that and I'll group it in with this PR.

**Review by bcardiff — APPROVED** [review]
I think is good to go. But since you add `BigDecimal#**(BigDecimal)`, aren't we also missing `BigDecimal#**(Float)`? It can be added later, just pointing out the lack of it.

Also, when I went to the integer division refactor I started leaving only the operations within the same type in the main file of the type, and moving to `big/number.cr` all the convenient operations to operate between different types. There is no need to change things on this, but I wanted to share the criteria.

**Comment by bcardiff** [comment]
Oh, please run the `$ crystal tool format ./src/big/big_decimal.cr` to make the CI happy. The formatter can be integrated with the editor of your preference.

**Comment by sugarfi** [comment]
Will do!

**Review by asterite — CHANGES_REQUESTED** [review]
Please revert the ** changes. Please keep PR simple. I already reviewed this, now I need to review it again, it's much more work for reviewers. It's much simpler to send one PR per thing to discuss/add.

**Comment by sugarfi** [comment]
Oh, my bad. Will revert!

**Comment by bcardiff** [comment]
@sugarfi with the force push I think you discarded the changes requested in https://github.com/crystal-lang/crystal/pull/9784#issuecomment-702504378

**Comment by sugarfi** [comment]
@bcardiff sorry for the delay, was caught up with other stuff. The specs should be re-added now. Thank you for your patience!

**Review by straight-shoota — APPROVED** [review]

**Comment by Sija** [comment]
This should be reopened, I guess.

**Comment by straight-shoota** [comment]
Hm, I don't think this work. The original repo and branch appear to have completely vanished. And GitHub even fails to reconstruct the commit contents.
Even the commits are not accessible in this PR. But they are still available on the repo directly: https://github.com/crystal-lang/crystal/tree/ca391dd4e19f91c819aeff9b57b5efaf8c69ea8b

**Comment by Blacksmoke16** [comment]
Closing in favor of #13255.

**Pull request closed without merging**